File: 18_정지윤/4949.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    private void solve() {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
            while (true) {
                String str = br.readLine();
                Stack<Character> stack = new Stack<>();
                boolean condition = true;

                if (str.equals(".")) {
                    break;
                }

                for (int i = 0; i < str.length(); ++i) {
                    if (str.charAt(i) == '[') {
                        stack.push(str.charAt(i));
                    } else if (str.charAt(i) == '(') {
                        stack.push(str.charAt(i));
                    } else if (str.charAt(i) == ']') {
                        if (!stack.empty()) {
                            if (stack.peek() == '[') {
                                stack.pop();
                            }else {
                                condition = false;
                            }
                        } else {
                            condition = false;
                        }
                    } else if (str.charAt(i) == ')') {
                        if (!stack.empty()) {
                            if (stack.peek() == '(') {
                                stack.pop();
                            }else {
                                condition = false;
                            }
                        }else {
                            condition = false;
                        }
                    }
                }

                if (stack.empty() && condition) {
                    System.out.println("yes");
                } else {
                    System.out.println("no");
                }

            }
        }catch (IOException e) {
            e.printStackTrace();
        }
    }
    public static void main(String[] args){
        new Main().solve();
    }
}

File: 19_조예린/4949.py
while True:
    s = input()
    if s == '.':
        break

    stack = []
    result = True
    for i in s:
        if i == '(' or i == '[':
            stack.append(i)
        elif i == ')':
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                result = False
                break
        elif i == ']':
            if stack and stack[-1] == '[':
                stack.pop()
            else:
                result = False
                break
    
    if not stack and result == True:
        print('yes')
    else:
        print('no')